Add batching for JLink

Tiwalun opened the pull request:
Support for batching SWD transfers for the JLink probe. 

Currently, the two transfers necessary for a read or write are batched,
but not multiple reads or writes.

The block read and write functionality is also implemented, which improves the RAM
read and write speed a lot.

Javier-varez commented:
This looks pretty good! There are a few things I missed in my implementation, this one is clearly more comprehensive. I was batching the transfers inside the driver JLink struct itself instead of providing an API for batched read/writes, but I see that this was already there for other probes. I also missed checking that the writes actually succeeded, not just checking that they were clocked in with a valid ack.

To me it would seem that letting the driver do the batching would improve the performance even further in most cases (combining single transactions that may not be combined otherwise). This is not possible for reads given that we return the value directly instead of a future, but It should be possible for writes (it is how I approached my implementation). Exposing then a flush function would still allow us to ensure that the operations completed up to a certain point whenever needed, so maybe it could be done as an improvement for the future.

Anyway, this PR certainly looks really good to me, I have no other comments.

Yatekii commented on probe-rs/src/probe/jlink/swd.rs:
```diff
@@ -0,0 +1,1240 @@
+    let mut read_index = 0;
+
+    for response_direction in expected_responses {
+        let response = parse_swd_response(&mut result[read_index..], response_direction);
```
Could you please fix this clippy complaint?

Yatekii commented on probe-rs/src/probe/jlink/swd.rs:
```diff
@@ -0,0 +1,1240 @@
+            "Unexpected response from target, does not conform to SWD specfication (ack={:?})",
+            ack
+        );
+        return Err(DapError::SwdProtocol);
```
Please fix this.

Yatekii reviewed (changes requested):
Thanks a lot for this!

I think it is very cleanly done with a lot of tests too :)

I would like to see the clippy lints in the newly introduced code fixed. I have pointed it out at two occasions but there is more I believe.

Also, You introduced a new `SwdProbe` trait. While I like this, it makes it impossible to use a JTag probe for ARM targets. Is that a usecase we never want to support? I understand that we do not support this right now.

Other than that, LGTM :)

Tiwalun commented:
I've renamed the `SwdProbe` trait to `DAPProbe`, to make it a bit clearer, but the new trait doesn't really change anything.

I moved the `DebugProbe + AsRef<dyn DebugProbe> + AsMut<dyn DebugProbe>` requirements from the `DAPAccess` trait to a new trait just to make it easier to test, but that doesn't change anything in regards to the functionality.

Yatekii reviewed (approved):
Looks & sounds good to me :) Thanks you for clarifying!

bors r+
